src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorFactory.cs

```diff
@@ -226,31 +226,132 @@ internal static IEnumerable<string> GetCommaSeparatedValues(string text)
+            string nameOrPrefix)
+        {
+            if (attributeNameOrPrefix == null)
+            {
+                // HtmlAttributeNameAttribute validates Name is non-null and non-empty. Both are valid for Prefix.
+                return true;
+            }
+
             // data-* attributes are explicitly not implemented by user agents and are not intended for use on
             // the server; therefore it's invalid for TagHelpers to bind to them.
-            if (attributeDescriptor.Name.StartsWith(DataDashPrefix, StringComparison.OrdinalIgnoreCase))
+            if (attributeNameOrPrefix.StartsWith(DataDashPrefix, StringComparison.OrdinalIgnoreCase))
             {
                 errorSink.OnError(
                     SourceLocation.Zero,
                     Resources.FormatTagHelperDescriptorFactory_InvalidBoundAttributeName(
-                        attributeDescriptor.PropertyName,
                         parentType.FullName,
+                        propertyName,
+                        nameOrPrefix,
+                        attributeNameOrPrefix,
                         DataDashPrefix));
 
                 return false;
             }
 
+            foreach (var character in attributeNameOrPrefix)
+            {
+                if (char.IsWhiteSpace(character) || InvalidNonWhitespaceNameCharacters.Contains(character))
+                {
+                    errorSink.OnError(
+                        SourceLocation.Zero,
+                        Resources.FormatTagHelperDescriptorFactory_InvalidBoundAttributeNameCharacter(
+                            parentType.FullName,
+                            propertyName,
+                            nameOrPrefix,
+                            attributeNameOrPrefix,
+                            character));
+
+                    return false;
+                }
+            }
+
             return true;
         }
 
-        private static TagHelperAttributeDescriptor ToAttributeDescriptor(PropertyInfo property)
+        private static TagHelperAttributeDescriptor ToAttributeDescriptor(
+            PropertyInfo property,
+            Type parentType,
+            ErrorSink errorSink)
         {
             var attributeNameAttribute = property.GetCustomAttribute<HtmlAttributeNameAttribute>(inherit: false);
-            var attributeName = attributeNameAttribute != null ?
-                                attributeNameAttribute.Name :
-                                ToHtmlCase(property.Name);
+            var attributeName = attributeNameAttribute?.Name ?? ToHtmlCase(property.Name);
+            var isStringProperty = property.PropertyType == typeof(string);
+            string prefix = null;
+            string objectCreationExpression = null;
+            string prefixedValueTypeName = null;
+            var areStringPrefixedValues = false;
+
+            var dictionaryTypeArguments = ExtractGenericInterface(property.PropertyType, typeof(IDictionary<,>))
```

Comment:
`ExtractGenericInterface` is only used for dictionary, no need for this method to be flexible.